lib/couchbase/protostellar/error_handling.rb

```diff
@@ -0,0 +1,144 @@
+# frozen_string_literal: true
+
+#  Copyright 2023. Couchbase, Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+require "couchbase/errors"
+require "google/rpc/error_details_pb"
+
+require_relative "request_behaviour"
+require_relative "retry/orchestrator"
+
+module Couchbase
+  module Protostellar
+    module ErrorHandling
+      TYPE_URL_PRECONDITION_FAILURE = "type.googleapis.com/google.rpc.PreconditionFailure"
+      TYPE_URL_RESOURCE_INFO = "type.googleapis.com/google.rpc.ResourceInfo"
+      TYPE_URL_ERROR_INFO = "type.googleapis.com/google.rpc.ErrorInfo"
+      TYPE_URL_BAD_REQUEST = "type.googleapis.com/google.rpc.BadRequest"
+
+      def self.handle_grpc_error(grpc_error, request)
```

Comment:
Added support for a few error types for now, the exact conversions rules have not yet been finalized